### PR TITLE
Fix gradient tensor mutate in `{adam/ftrl/rmprop/rmspropalex}_update`.

### PR DESCRIPTION
**Pull request by kshitij12345**
## Description ##
Rescaling the gradient used to update the data of `grad` passed as input.
Detailed in #15759 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fix the bug.
- [x] Relevant test.

### CODE REVIEW

File: src/operator/optimizer_op.cu [hunk 1]
```diff
@@ -261,8 +261,7 @@ NNVM_REGISTER_OP(ftml_update)
 .set_attr<FCompute>("FCompute<gpu>", FTMLUpdate<gpu>);
 
 NNVM_REGISTER_OP(adam_update)
-.set_attr<FCompute>("FCompute<gpu>", AdamUpdate<gpu>)
-.set_attr<FComputeEx>("FComputeEx<gpu>", AdamUpdateEx<gpu>);
```
Comment (sxjscience) [inline src/operator/optimizer_op.cu]: @eric-haibin-lin Is it safe to remove `AdamUpdateEx`?

Comment (kshitij12345) [reply src/operator/optimizer_op.cu]: Oops. I don't think we should remove it. Had removed it temporarily. Will undo it.

Thanks.

**Comment by kshitij12345** [comment]
@sxjscience @eric-haibin-lin @apeforest @larroy Could you please have a look. 

All the original tests pass and have added test to check only expected variables are mutated.

File: src/operator/optimizer_op-inl.h [hunk 2]
```diff
@@ -1596,57 +1611,76 @@ struct RMSPropAlexParam : public dmlc::Parameter<RMSPropAlexParam> {
...
+    const DType* weight_data, const DType* grad_data,
+    const DType clip_gradient, const DType rescale_grad,
+    const DType gamma1, const DType gamma2,
+    const DType lr, const DType wd,
+    const DType clip_weights, const DType epsilon,
+    const OpReqType req) {
+    using namespace mshadow_op;
+
+    const DType rescaled_grad = rescale_grad * grad_data[i] +
+           wd * weight_data[i];
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply src/operator/optimizer_op-inl.h]: @sxjscience Done.

Thanks for the suggestions. Makes if more readable.

**Comment by Vikas-kum** [comment]
@kshitij12345 @apeforest Test is failing for this change - 
export MXNET_TEST_SEED=412298777 
nosetests -v tests/python/unittest/test_ndarray.py:test_update_ops_mutation

http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-cpu/detail/master/1039/pipeline

**Comment by kshitij12345** [comment]
@Vikas-kum Thanks.

I have checked and found that the difference is -5.9604645e-08 which is lower than the default `rtol` of `1e-07` for assert_mutate.
It works as expected with more sensitive tolerance.
Will send a  PR with more sensitive tolerance.

**Comment by ChaiBapchya** [comment]
So I tested *_update ops and it turns out passing randn (which samples from uniform distribution) to *_update op (e.g. adam_update) gives output that may consist nans

```
> mx.nd.adam_update(mx.nd.random.randn(10),mx.nd.random.randn(10),mx.nd.random.randn(10),mx.nd.random.randn(10),.01)

[        nan         nan  0.09059371         nan -0.1433481          nan
 -0.5170411   0.381852    0.62985003         nan]
```
Now the way you've tested is checked if input and output has mutation after *_update method is called. Does that take into consideration the NaNs?

@kshitij12345

**Comment by kshitij12345** [comment]
@ChaiBapchya 
Thanks for checking. Interesting find. Haven't considered this case and thus there is no code for handling this condition.
Will look into this.

**Comment by ChaiBapchya** [comment]
Sure. I think it needs to be handled.

**Pull request merged**